File: SecureBlackbox Samples/JWC Signer/python/jwc_signer.py
# 
# SecureBlackbox 2022 Python Edition - Sample Project
# 
# This sample project demonstrates the usage of SecureBlackbox in a 
# simple, straightforward way. It is not intended to be a complete 
# application. Error handling and other checks are simplified for clarity.
# 
# www.nsoftware.com/secureblackbox
# 
# This code is subject to the terms and conditions specified in the 
# corresponding product license agreement which outlines the authorized 
# usage and restrictions.
# 

import sys
import string
from secureblackbox import *

input = sys.hexversion<0x03000000 and raw_input or input

def ensureArg(args, prompt, index):
  if len(args) <= index:
    while len(args) <= index:
      args.append(None)
    args[index] = input(prompt)
  elif args[index] == None:
    args[index] = input(prompt)

def displayHelp():
    print(
        "NAME\n"
        "  jwsigner -- SecureBlackbox SymmetricCrypto Demo Application\n\n"
        "SYNOPSIS\n"
        "  jwsigner -s/-v [-input input_data] [-sig signature_data] [-cert certificate_file] [-certpass certificate_password] [-compact]\n\n"
        "DESCRIPTION\n"
        "  This sample illustrates how to create a detached signature over a text string.\n"
        "  Used to sign and verify data.\n\n"
        "  The options are as follows:\n\n"
        "  -s            Whether to sign input data. \n\n"
        "  -v            Whether to verify signature data. \n\n"
        "  -input        An input data to sign/verify (Required). \n\n"
        "  -sig          An signature data to verify (Required to verify). \n\n"
        "  -cert         The certificate used to encrypt file (Required). \n\n"
        "  -certpass     The password for the certificate. \n\n"
        "  -compact      Whether to use compact format \n\n"
        "EXAMPLES\n"
        "	jwsigner -s -input \"And now that you don’t have to be perfect, you can be good.\" -cert C:\\\\certs\\\\mycert.pfx -certpass mypassword \n\n"
        "	jwsigner -v -input \"And now that you don’t have to be perfect, you can be good.\" \n"
        "		-sig eyJhbGciOiJkaXIiLCJlbmMiOiJBMjU2Q0JDLUhTNTEyIn0..kuN2U -cert C:\\\\certs\\\\mycert.pfx -certpass mypassword -compact \n\n"
    )

if (len(sys.argv) < 6):
    displayHelp()
    sys.exit(1)
else:
    crypto = PublicKeyCrypto()
    cm = CertificateManager()
    ckm = CryptoKeyManager()
    
    sign = False
    verify = False
    inputD = ""
    signatureD = ""
    certF = ""
    certpass = ""
    compact = False
    
    for x in range(len(sys.argv)):
        if (sys.argv[x].startswith("-")):
            if (sys.argv[x].lower() == "-s"):
                sign = True
            if (sys.argv[x].lower() == "-v"):
                verify = True
            if (sys.argv[x].lower() == "-input"):
                inputD = sys.argv[x+1]
            if (sys.argv[x].lower() == "-sig"):
                signatureD = sys.argv[x+1]
            if (sys.argv[x].lower() == "-cert"):
                certF = sys.argv[x+1]
            if (sys.argv[x].lower() == "-certpass"):
                certpass = sys.argv[x+1]
            if (sys.argv[x].lower() == "-compact"):
                compact = True

    if (not (sign or verify)):
        print("-s or -v is required.\n")
        displayHelp()
        sys.exit(1)

    if (sign and verify):
        print("-Use only one -s or -v parameter.\n")
        displayHelp()
        sys.exit(1)

    if (inputD == ""):
        print("-input is required.\n")
        displayHelp()
        sys.exit(1)
        
    if (verify and (signatureD == "")):
        print("-signature is required.\n")
        displayHelp()
        sys.exit(1)
    
    if (certF == ""):
        print("-cert is required.\n")
        displayHelp()
        sys.exit(1)
            
    try:
        cm.import_from_file(certF, certpass)
        ckm.set_cert_handle(cm.get_cert_handle())
        ckm.import_from_cert()
        crypto.set_key_handle(ckm.get_key_handle())
      
        inputB = inputD.encode('utf-8')
        
        if sign:
            if compact:
                crypto.set_output_encoding(3); # cetCompact
            else:
                crypto.set_output_encoding(4); # cetJSON
                
            signatureB = crypto.sign(inputB, True)         
            print("Signature: " + signatureB.decode('utf-8', errors='backslashreplace'))
        else:
            if compact:
                crypto.set_input_encoding(3); # cetCompact
            else:
                crypto.set_input_encoding(4); # cetJSON
                
            signatureB = signatureD.encode('utf-8')
            
            crypto.verify_detached(inputB, signatureB)

            if (crypto.get_signature_validation_result == svtSignerNotFound):
                print("Signer not found")
            elif (crypto.get_signature_validation_result == svtFailure):
                print("Signature verification failed")
            elif (crypto.get_signature_validation_result == svtCorrupted):
                print("Signature is invalid")
            elif (crypto.get_signature_validation_result == svtCorrupted):
                print("Signature validated successfully.")
            else:
                print("Verification unknown.") 
    except Exception as e: 
        print(e)

File: SecureBlackbox Samples/JWC Signer/php/jwc_signer.php
<?php
/*
 * SecureBlackbox 2024 PHP Edition - Sample Project
 *
 * This sample project demonstrates the usage of SecureBlackbox in a 
 * simple, straightforward way. It is not intended to be a complete 
 * application. Error handling and other checks are simplified for clarity.
 *
 * www.nsoftware.com/secureblackbox
 *
 * This code is subject to the terms and conditions specified in the 
 * corresponding product license agreement which outlines the authorized 
 * usage and restrictions.
 */
require_once('../include/secureblackbox_publickeycrypto.php');
require_once('../include/secureblackbox_cryptokeymanager.php');
require_once('../include/secureblackbox_certificatemanager.php');
require_once('../include/secureblackbox_const.php');
?>
<style>
table { width: 100% !important; }
td { white-space: nowrap; }
td input { width: 100%; }
td:last-child { width: 100%; }
</style>

<div width="90%">
  <form method=POST>
    <h2>Public key crypto Demo</h2>
    
    <h3>General Options</h3>
    <label><input type ="radio" checked="checked" name="comtype" value="Sign" /> Sign </label>
    <label><input type ="radio" name="comtype" value="Verify" /> Verify </label>
    <br/>
    <br/>

    <table>
      <tr><td>Input:</td><td><input type=text name=inputData value=""></td></tr>
      <tr><td>Signature (for verifying):</td><td><input type=text name=signatureData value=""></td></tr>
    </table>
    <br/>
    <br/>

    <label> <input type="checkbox" name="compact" value="1" size="25"/> Compact </label>
    <br/>
    <br/>

    <b>Signing Certificate</b>
    <table>
      <tr><td>Certificate File:</td><td><input type=text name=sCertFile value=""></td></tr>
      <tr><td>Password:</td><td><input type=password name=sCertPass value=""></td></tr>
    </table>
    <br/>
    <br/>

    <input type="submit" value="Sign/Verify" />
  </form>
</div><br/>

<?php
  if ($_SERVER['REQUEST_METHOD'] == "POST") {
    $crypto = new SecureBlackbox_PublicKeyCrypto();
    $certmgr = new SecureBlackBox_CertificateManager();
    $keymgr = new SecureBlackBox_CryptoKeyManager();
    
    try {
      // General options
      $inputData = $_REQUEST['inputData'];
      $signatureData = $_REQUEST['signatureData'];
      $compact = (isset($_REQUEST['compact']) && (($_REQUEST['compact'] == 'yes') || ($_REQUEST['compact'] == '1')));

      // Signing options
      $certmgr->doImportFromFile($_REQUEST['sCertFile'], $_REQUEST['sCertPass']);
      $keymgr->setCertHandle($certmgr->getCertHandle());
      $keymgr->doImportFromCert();
      $crypto->setKeyHandle($keymgr->getKeyHandle());

      $comtype = $_POST["comtype"];

      if ($comtype == "Sign")
      {
        if ($compact)
        {
          $crypto->setOutputEncoding(PUBLICKEYCRYPTO_OUTPUTENCODING_COMPACT);
        }
        else
        {
          $crypto->setOutputEncoding(PUBLICKEYCRYPTO_OUTPUTENCODING_JSON);
        }

        $signatureData = $crypto->doSign($inputData, TRUE);
        echo "<h2>Signature string: " . $signatureData . "</h2>";
      }
      else
      {
        if ($compact)
        {
          $crypto->setInputEncoding(PUBLICKEYCRYPTO_INPUTENCODING_COMPACT);
        }
        else
        {
          $crypto->setInputEncoding(PUBLICKEYCRYPTO_INPUTENCODING_JSON);
        }

        $crypto->doVerifyDetached($inputData, $signatureData);

        switch($crypto->getSignatureValidationResult())
        {
          case PUBLICKEYCRYPTO_SIGNATUREVALIDATIONRESULT_VALID:  echo "<h2>Signature validated successfully</h2>";          break;
          case PUBLICKEYCRYPTO_SIGNATUREVALIDATIONRESULT_CORRUPTED:  echo "<h2>Signature is invalid</h2>";      break;
          case PUBLICKEYCRYPTO_SIGNATUREVALIDATIONRESULT_SIGNER_NOT_FOUND:  echo "<h2>Signer not found</h2>"; break;
          case PUBLICKEYCRYPTO_SIGNATUREVALIDATIONRESULT_FAILURE:  echo "<h2>Signature verification failed</h2>";        break;
          default: echo "<h2>Unknown</h2>";        break;
        }
      }
    }
    catch (exception $e) {
      echo "<h2>Signing Failure (Details Below)</h2><p>" . $e->getMessage() . "</p>";
    }
  }
?>